buildpack.md

```diff
@@ -1135,6 +1137,39 @@ Each `key`:
 ## Deprecations
 This section describes all the features that are deprecated.
 
+### buildpack.toml (TOML) `stacks` Array
+
+_Deprecated in Buildpack API 0.10._
+
+The `stacks` array is deprecated.
+
+```toml
+[[stacks]]
+id = "<stack ID>"
+mixins = ["<mixin name>"]
+```
+
+Each stack in `stacks` either:
+- MUST identify a compatible stack:
+    - `id` MUST be set to a [valid stack ID](https://github.com/buildpacks/spec/blob/main/platform.md#stack-id).
+    - `mixins` MAY contain one or more mixin names.
+- Or MUST indicate compatibility with any stack:
+    - `id` MUST be set to the special value `"*"`.
+    - `mixins` MUST be empty.
+
+If an `order` is specified, then `stacks` MUST NOT be specified.
+
+Tools reading `buildpack.toml` will translate any section that sets `stacks.id = "io.buildpacks.stacks.bionic` to:
```

Comment:
```suggestion
Tools reading `buildpack.toml` will translate any section that sets `stacks.id = "io.buildpacks.stacks.bionic"` to:
```

Reply:
> if we have an older buildpack on a base image that doesn't have stack Id

Base image authors are encouraged to provide both the old and the new formats during the migration period. From the RFC:

> All existing labels and environment variables for stacks and mixins may be preserved on a base image until all users have migrated to the new format.

> Vice versa what about other stack ids apart from bionic?

For newer buildpack on an older base image, we can infer the target data for `io.buildpacks.stacks.bionic` (this is mentioned in the platform spec PR [here](https://github.com/buildpacks/spec/pull/335/files#diff-e603760990971da3f77be4bb8d77c3405098f006814fd8c054d2d15f395b8330R1067-R1072)) - but for other stack IDs, I guess it would just fail.

> Should we just ensure that the stack author needs to define a stack ID if they are using an older buildpack api?

A warning during `pack builder create` might be appropriate.

Reply:
The stack author doesn't actually know or declare what APIs are being used right? Even for a builder looking at the [builder docs](https://buildpacks.io/docs/reference/config/builder-config/), there's no specification of which APIs are being used.

Reply:
Is the migration path that both Base Images AND Buildpacks need to support both paths? A base image may not know all the buildpacks running on it and a buildpack may not know what base images it runs on.

Reply:
> What warning are you thinking of here?

I went back and read my comment, and I'm not sure :/ 

For `pack builder create`, we pull down and unpack each buildpack referenced in builder.toml to validate that the stacks required by the buildpack in buildpack.toml are satisfied by the base image. I think we still want to do this validation, so there is only pass & fail state when older buildpacks are included in the builder.

Reply:
Yeah, I think it's much simpler when you think about buildpacks in a builder. It gets much more complicated in a PaaS space where a builder is really a base image + lifecycle and buildpacks could come from anywhere or vice versa if a Buildpack can expect to run in a multitude of Builders. In think there's a 2x2 matrix we have to deal with:

* Buildpack (Stack) <-> Builder (Stack)
* Buildpack (Stackless) <-> Builder (Stackless), everyone migrates success!
* Buildpack (Stackless) <-> Builder (Stack), Buildpack needs to support `CNB_STACK_ID`
* Buildpack (Stack) <-> Builder (Stackless), while the Builder has upgraded it needs to support `CNB_STACK_ID` to pass to the Buildpack

I'm not sure where the best place to throw deprecation warnings, since for maximum compatibility Buildpacks and Builders need to continue to support `CNB_STACK_ID`. I do think this is something we should drop when `1.0` lands of the specs.

Reply:
This may be summarized in the [table](https://github.com/buildpacks/spec/pull/336#issuecomment-1454099934) below.

Personally I believe the deprecation steps outlined there (in the green cells) are enough. The lifecycle can emit warnings whenever it has to derive stuff that should be present according to the Buildpack and/or Platform API version.

Reply:
@natalieparellano in that table, will stack be required to be supplied?

Reply:
Discussed in 3/9/23 Working Group, along with this (newer, better) table to summarize the backwards compat strategy: https://docs.google.com/spreadsheets/d/1_lUDh8V_A6SHV9w2ioGwqAu821WWaItSYiUKqrlmr_w/edit#gid=240821534

We think that reading values from /etc/os-release will fill the last remaining gaps.